### ISSUE TITLE
Add a node-view-counter function?

### ISSUE THREAD

**Issue by docwilmot** [backdrop-contrib/forum#9]
Most common forums have a "Number of times this post has been viewed" or equivalent. In D7 this could use the core Statistics module to do the counting, but that has been removed from Backdrop. We could implement a simple counter in Forum by adding a column to the 'forum' table for number of node views and incrementing on `hook_exit()`. Not sure about the performance implications of this though. Anyone think its worth it?

**Comment by klonos** [comment]
Would this count unique views though? I think it should.

**Comment by docwilmot** [comment]
> Would this count unique views though? I think it should.

That would make it more complicated, as we'd need to store every IP that's viewed every forum node, meaning another table. 

And I'm not sure its necessary. If fewer people are viewing a really good discussion 20 times a day that still means it is "hotter"than a view with a catchy title that lots of people look at once then never again. Justifiable argument? Not sure, just saying.

Perhaps if/when someone ports Statistics we can add a dependency to it. 

Would be great to get an expert opinion on the performance implications of adding a counter which does a `db_merge` on each node view for a really popular forum. @quicksketch @serundeputy @Gormartsen @jenlampton. I'm thinking just a variation of the following from Statistics:

```
function forum_exit()
// Check is a node page and forum type then do:
      db_merge('forum')
        ->key(array('nid' => arg(1)))
        ->fields(array(
          'views' => 1,
        ))
        ->expression('views', 'views+ 1')
        ->execute();
```

**Comment by jenlampton** [comment]
I don't think tracking views in the db is a good idea. Statistics counters are a job best left to javascript, and even better left to the experts (like Google Analytics). I would rather see some integration with GA than a port of the statistics module, as the data is likely to be more accurate (and won't crash the site!) :)

**Comment by klonos** [comment]
...if the views counter is incremented on every page load, then people could potentially refresh the page a few dozen times to make it "hot". Just saying.

Besides, I was talking about views counting registered users, so I thought that the counter would increase by user ID, not IP.

**Comment by docwilmot** [comment]
> people could potentially refresh the page a few dozen times to make it "hot"

True but best accommodate the unscrupulous than hit performance I think.

> counter would increase by user ID, not IP.

It would need be IP to account for tracking anonymous users

> I would rather see some integration with GA

The ported GA wouldn't do, it only collects stats; we'd need a module that integrates with the GA API to **retrieve** stats, of which currently none have been ported.

But @jenlampton although I agree GA integration is ideal, is it reasonable to include this as an **option** for those who may not want to install another module? So a user can either:
- disable counting
- use the poor-mans-counting code above
- install a GA API module.

How much of a "bad thing" is that code snippet up there likely to be if it only counts forum node page views? I suspect Statistics module sucked because it counted every view of every page and a whole lot of other stuff as well. Note that I'm not arguing, but genuinely asking, because I don't know.

**Comment by jenlampton** [comment]
> I suspect Statistics module sucked because it counted every view of every page and a whole lot of other stuff as well. Note that I'm not arguing, but genuinely asking, because I don't know.

In order to prevent sites from falling over, statistics module truncated the data stored based on time intervals. If we followed the same pattern, we'd only get the # of views in the last 30 days (or similar). I'm not sure that's enough data for our needs here.

There was a way to disable the auto-deletion of page-view-counts over time, but If you use that feature and your site has any reasonable number of users (mine all had far less than backdropcms.org will) the site would very soon end up going very slowly.

Not only was statistics module doing something terribly slow (database write) on an action that should be terribly fast (database read), but it also stored each "view" as an individual record in the database (along with referrer, IP, and other useful statistics data) but all those records also made the extraction of the view "count" painful as well.

What if we have 1000 people viewing the same page at the same time? That page can be served from cache at least 999 times, but there's no way to streamline 1000 individual database writes. That's one fast read, and 1000 slow writes!

Also, how do we account for bounces if we are recording all the page views? I read a lot of useless forum entries on drupal.org, stack exchange, etc, but I have to read them first to know if they are useless or not. I wouldn't want that read action bumping up a useless forum post to the top of the list for other people, so now they have to read it too.

Can we implement something more useful like a thumbs-up style "like" action, that would allow people to decide if their "views" should be counted, instead?

**Comment by klonos** [comment]
> > counter would increase by user ID, not IP.
> 
> It would need be IP to account for tracking anonymous users

Yeah, I was thinking something like this: if user is logged in (thus registered), then check if user viewed the node previously (we need this to mark content "unread" for that user anyways). If yes, do nothing. If no, counter++. If user is not logged in (anonymous), then check if IP has viewed the content previously. If yes, do nothing. If no, counter++.

Does that make sense?

**Comment by docwilmot** [comment]
Thanks @klonos, I understand. I've just discovered there's a `history` table in core which saves when last a person looked at a node. So for registered users, then we can in fact check that table for previous views. BUT, I'm still uncertain about allowing only one view to be counted. Google says a pageview happens on each refresh, understandable; but even a `unique pageview` is based on the session:

> A unique pageview, as seen in the Content Overview report, aggregates pageviews that are generated by the same user during the same session. A unique pageview represents the number of sessions during which that page was viewed one or more times.

A Backdrop session lasts 23 days. Is that reasonable? Or could we use some configurable logic here, like an admin setting for what time interval to consider a new pageview unique?

Incidentally, though I'm still pursuing this, I've ported the Statistics module from D7 core instead, so all the pageview tracking will be in there instead.
